### DISCUSSION TITLE
Haskell-Tools Not Loading

### DISCUSSION THREAD

**Discussion by TorNATO-PRO** [LazyVim/LazyVim#705] [Q&A] [answered]
I was wondering if anyone has a lazy.nvim configuration for Haskell.

## Problem

I am running into the following problem (likely with my config)

```
E5108: Error executing lua [string ":lua"]:1: module 'haskell-tools' not found:
        no field package.preload['haskell-tools']
cache_loader: module haskell-tools not found
cache_loader_lib: module haskell-tools not found
```

## What I've Tried

I have the following configuration for NvChad (which uses lazy.nvim for plugin management).

```lua
local plugins = {
  {
    "nvim-treesitter/nvim-treesitter",
    opts = {
      ensure_installed = { "html", "css", "bash", "haskell", "c", "python" }
    },
  },
  {
    "mrcjkb/haskell-tools.nvim",
    dependencies = {
      "nvim-lua/plenary.nvim",
      "nvim-telescope/telescope.nvim",
    },
    config = function()
      require "custom.configs.haskell"
    end,
    branch = "1.x.x",
    lazy = false,
  },
}

return plugins
```

Which depends on the following from [haskell-tools](https://github.com/mrcjkb/haskell-tools.nvim) README. 

```lua
local ht = require('haskell-tools')
local def_opts = { noremap = true, silent = true, }
ht.start_or_attach {
  hls = {
    on_attach = function(client, bufnr)
      local opts = vim.tbl_extend('keep', def_opts, { buffer = bufnr, })
      -- haskell-language-server relies heavily on codeLenses,
      -- so auto-refresh (see advanced configuration) is enabled by default
      vim.keymap.set('n', '<space>ca', vim.lsp.codelens.run, opts)
      vim.keymap.set('n', '<space>hs', ht.hoogle.hoogle_signature, opts)
      vim.keymap.set('n', '<space>ea', ht.lsp.buf_eval_all, opts)
    end,
  },
}

-- Suggested keymaps that do not depend on haskell-language-server:
local bufnr = vim.api.nvim_get_current_buf()
-- set buffer = bufnr in ftplugin/haskell.lua
local opts = { noremap = true, silent = true, buffer = bufnr }

-- Toggle a GHCi repl for the current package
vim.keymap.set('n', '<leader>rr', ht.repl.toggle, opts)
-- Toggle a GHCi repl for the current buffer
vim.keymap.set('n', '<leader>rf', function()
  ht.repl.toggle(vim.api.nvim_buf_get_name(0))
end, def_opts)
vim.keymap.set('n', '<leader>rq', ht.repl.quit, opts)

-- Detect nvim-dap launch configurations
-- (requires nvim-dap and haskell-debug-adapter)
ht.dap.discover_configurations(bufnr)
```

I appreciate any help - thanks!

**Comment by funarog** [comment]
I too am having issues trying to figure this out. The examples and documentation are not as clear or intuitive as folke believes it is. how  And I am senior programmer. Which means I can eventually  figure this out, but not sure it is worth the effort.

**Comment by funarog** [comment]
The following appears to be working for the most part. I am unsure how well it follows the lazy.nvim spec. I am going through the keymaps and checking out functionality.  Create the a file in your ~/.config structure  ../lua/plugins/haskell_tools.lua and add these contents: 
---------------
`return {
  -- add haskell tools
  "mrcjkb/haskell-tools.nvim",
  name = "haskell-tools",
  dev = false,
  lazy = true,
  dependencies = {
    "nvim-lua/plenary.nvim",
    "nvim-telescope/telescope.nvim", -- optional
  },
  init = function() --function called at startup ??GVF nvim startup versus config
    local ht = require("haskell-tools")
    local def_opts = { noremap = true, silent = true }
    ht.start_or_attach({
      hls = {
        on_attach = function(_, bufnr)
          local opts = vim.tbl_extend("keep", def_opts, { buffer = bufnr })
          -- haskell-language-server relies heavily on codeLenses,
          -- so auto-refresh (see advanced configuration) is enabled by default
          vim.keymap.set("n", "<space>ca", vim.lsp.codelens.run, opts)
          vim.keymap.set("n", "<space>hs", ht.hoogle.hoogle_signature, opts)
          vim.keymap.set("n", "<space>ea", ht.lsp.buf_eval_all, opts)
        end,
      },
    })
    -- Suggested keymaps that do not depend on haskell-language-server:
    local bufnr = vim.api.nvim_get_current_buf()
    -- set buffer = bufnr in ftplugin/haskell.lua
    local opts = { noremap = true, silent = true, buffer = bufnr }

    -- Toggle a GHCi repl for the current package
    vim.keymap.set("n", "<leader>rr", ht.repl.toggle, opts)
    -- Toggle a GHCi repl for the current buffer
    vim.keymap.set("n", "<leader>rf", function()
      ht.repl.toggle(vim.api.nvim_buf_get_name(0))
    end, def_opts)
    vim.keymap.set("n", "<leader>rq", ht.repl.quit, opts)

    -- Detect nvim-dap launch configurations
    -- (requires nvim-dap and haskell-debug-adapter)
    ht.dap.discover_configurations(bufnr)
  end,
  -- opts = { noremap = true, silent = true, buffer = bufnr },

  -- config executed when the plug in loads ??GVF this is very unclear
  -- main
  -- build
  branch = "1.x.x", --recommended
  -- tag
  -- commit
  -- version
  pin = false, -- when true  this p[lug in will be included in updates]
  -- submodules
  -- event
  -- cmd
  ft = "haskell",
  -- keys = {}
  -- module
  -- priority
}
`

**Comment by TorNATO-PRO** [comment, marked as answer]
@funarog Thanks for the advice. I figure that the problem might be with NvChad so I removed that and am using base `lazy.nvim`.

```
λ ~/.config/nvim/lua$ tree
.
├── configs
│   └── haskell.lua
└── plugins.lua

1 directory, 2 files
```

```lua
-- config.lua
return {
    {
        "folke/tokyonight.nvim",
        lazy = false,
        priority = 1000,
        config = function() vim.cmd([[colorscheme tokyonight]]) end
    }, {
        "mrcjkb/haskell-tools.nvim",
        dependencies = {
            "nvim-lua/plenary.nvim", "nvim-telescope/telescope.nvim"
        },
        branch = "1.x.x",
	ft = "haskell",
        config = function() 
		require("configs.haskell")
	end
    },
    { "mfussenegger/nvim-dap" },
}
```

```lua
-- haskell.lua
require('haskell-tools').start_or_attach {
  tools = { -- haskell-tools options
    codeLens = {
      -- Whether to automatically display/refresh codeLenses
      -- (explicitly set to false to disable)
      autoRefresh = true,
    },
    hoogle = {
      -- 'auto': Choose a mode automatically, based on what is available.
      -- 'telescope-local': Force use of a local installation.
      -- 'telescope-web': The online version (depends on curl).
      -- 'browser': Open hoogle search in the default browser.
      mode = 'auto',
    },
    hover = {
      -- Whether to disable haskell-tools hover
      -- and use the builtin lsp's default handler
      disable = false,
      -- Set to nil to disable
      border = {
        { '╭', 'FloatBorder' },
        { '─', 'FloatBorder' },
        { '╮', 'FloatBorder' },
        { '│', 'FloatBorder' },
        { '╯', 'FloatBorder' },
        { '─', 'FloatBorder' },
        { '╰', 'FloatBorder' },
        { '│', 'FloatBorder' },
      },
      -- Stylize markdown (the builtin lsp's default behaviour).
      -- Setting this option to false sets the file type to markdown and enables
      -- Treesitter syntax highligting for Haskell snippets
      -- if nvim-treesitter is installed
      stylize_markdown = false,
      -- Whether to automatically switch to the hover window
      auto_focus = false,
    },
    definition = {
      -- Configure vim.lsp.definition to fall back to hoogle search
      -- (does not affect vim.lsp.tagfunc)
      hoogle_signature_fallback = false,
    },
    repl = {
      -- 'builtin': Use the simple builtin repl
      -- 'toggleterm': Use akinsho/toggleterm.nvim
      handler = 'builtin',
      builtin = {
        create_repl_window = function(view)
          -- create_repl_split | create_repl_vsplit | create_repl_tabnew | create_repl_cur_win
          return view.create_repl_split { size = vim.o.lines / 3 }
        end
      },
      -- Can be overriden to either `true` or `false`.
      -- The default behaviour depends on the handler.
      auto_focus = nil,
    },
    -- Set up autocmds to generate tags (using fast-tags)
    -- e.g. so that `vim.lsp.tagfunc` can fall back to Haskell tags
    tags = {
      enable = vim.fn.executable('fast-tags') == 1,
      -- Events to trigger package tag generation
      package_events = { 'BufWritePost' },
    },
    dap = {
      cmd = { 'haskell-debug-adapter' },
    },
  },
  hls = { -- LSP client options
    -- ...
    default_settings = {
      haskell = { -- haskell-language-server options
        formattingProvider = 'ormolu',
        -- Setting this to true could have a performance impact on large mono repos.
        checkProject = true,
        -- ...
      }
    }
  }
}
```


I hope that someone with the same problem can find this useful in the future.

**Reply by blackmirroxx** [reply]
Well technically it works but, for every single keystroke I do -> My screen gets flooded with error messages. 
``
I get a long never ending error message: 
Error detected while processing BufReadPost Autocommands for "*":                                                     │ 9880  ./configure --with-lua-include=/usr/local/include
Error executing lua callback: /usr/local/share/nvim/runtime/filetype.lua:21: Error executing lua: /usr/local/share/nvi│ 9881  find / -type f -name lua.h
m/runtime/filetype.lua:22: BufReadPost Autocommands for "*"..FileType Autocommands for "*"..function <SNR>1_LoadFTPlug│ 9882  ./configure --with-lua-include=/usr/local/include
in[19]..script /home/l8star/.config/nvim/ftplugin/haskell.lua: Vim(runtime):E5113: Error while calling lua chunk: /hom│ 9883  ./configure --with-lua-include=/usr/include/lua5.1/lua.h
e/l8star/.config/nvim/ftplugin/haskell.lua:2: module 'haskell-tools' not found:                                       │ 9884  ./configure --with-lua-include=/usr/include/lua5.1/
        no field package.preload['haskell-tools']                                                                     │ 9885  make -j02
cache_loader: module haskell-tools not found                                                                          │ 9886  pwd
cache_loader_lib: module haskell-tools not found                                                                      │ 9887  sh -c 'curl -fLo "${XDG_DATA_HOME:-$HOME/.local/share}"/nvim/site/autoload/plug.vim --create-dirs \\n       htt
        no file './haskell-tools.lua'                                                                                 │ps://raw.githubusercontent.com/junegunn/vim-plug/master/plug.vim'
        no file '/home/l8star/neovim/neovim/.deps/usr/share/luajit-2.1.0-beta3/haskell-tools.lua'                     │ 9888  ls
        no file '/usr/local/share/lua/5.1/haskell-tools.lua'                                                          │ 9889  Plug 'nvim-lua/plenary.nvim'
        no file '/usr/local/share/lua/5.1/haskell-tools/init.lua'                                                     │ 9890  sh -c 'curl -fLo "${XDG_DATA_HOME:-$HOME/.local/share}"/nvim/site/autoload/plug.vim --create-dirs \\n       htt
        no file '/home/l8star/neovim/neovim/.deps/usr/share/lua/5.1/haskell-tools.lua'                                │ps://raw.githubusercontent.com/junegunn/vim-plug/master/plug.vim'
        no file '/home/l8star/neovim/neovim/.deps/usr/share/lua/5.1/haskell-tools/init.lua'                           │ 9891* nvim init.lua
        no file './haskell-tools.so'                                                                                  │ 9892  curl -fLo ~/.var/app/io.neovim.nvim/data/nvim/site/autoload/plug.vim --create-dirs \\n    https://raw.githubuse
        no file '/usr/local/lib/lua/5.1/haskell-tools.so'                                                             │rcontent.com/junegunn/vim-plug/master/plug.vim
        no file '/home/l8star/neovim/neovim/.deps/usr/lib/lua/5.1/haskell-tools.so'                                   │ 9893  ls
        no file '/usr/local/lib/lua/5.1/loadall.so'                                                                   │ 9894  history
stack traceback:                                                                                                      │ 9895  luarocks install plenary.nvim
        [C]: in function 'require'                                                                                    │λ ~/luaRocks/luarocks-3.8.0/ luarocks install haskell-tools.nvim
        /home/l8star/.config/nvim/ftplugin/haskell.lua:2: in main chunk                                               │Installing https://luarocks.org/haskell-tools.nvim-1.10.2-1.src.rock
        [C]: in function 'nvim_cmd'                                                                                   │                                                                                                                              /usr/local/share/nvim/runtime/filetype.lua:22: in function </usr/local/share/nvim/runtime/filetype.lua:21>    │haskell-tools.nvim 1.10.2-1 depends on lua >= 5.1 (5.1-1 provided by VM)
        [C]: in function 'nvim_buf_call'                                                                              │haskell-tools.nvim 1.10.2-1 depends on plenary.nvim (scm-1 installed)
        /usr/local/share/nvim/runtime/filetype.lua:21: in function </usr/local/share/nvim/runtime/filetype.lua:10>    │haskell-tools.nvim 1.10.2-1 is now installed in /home/l8star/luaRocks/luarocks-3.8.0/./lua_modules (license: GPL-2.0)
stack traceback:                                                                                                      │
        [C]: in function 'nvim_cmd'                                                                                   │λ ~/luaRocks/luarocks-3.8.0/ luarocks install haskell-tools.nvim                                                              /usr/local/share/nvim/runtime/filetype.lua:22: in function </usr/local/share/nvim/runtime/filetype.lua:21>    │Installing https://luarocks.org/haskell-tools.nvim-1.10.2-1.src.rock
        [C]: in function 'nvim_buf_call'                                                                              │
        /usr/local/share/nvim/runtime/filetype.lua:21: in function </usr/local/share/nvim/runtime/filetype.lua:10>    │haskell-tools.nvim 1.10.2-1 depends on lua >= 5.1 (5.1-1 provided by VM)
stack traceback:                                                                                                      │haskell-tools.nvim 1.10.2-1 depends on plenary.nvim (scm-1 installed)                                                         [C]: in function 'nvim_buf_call'                                                                              │haskell-tools.nvim 1.10.2-1 is now installed in /home/l8star/luaRocks/luarocks-3.8.0/./lua_modules (license: GPL-2.0)
        /usr/local/share/nvim/runtime/filetype.lua:21: in function </usr/local/share/nvim/runtime/filetype.lua:10> 
``

Technically it works -> as long as just read 
Parts of my packages are installed with cabal. Other parts are installed with emerge (in gentoo). It could be that there are some version conflicts going on.
